packages/clay-date-picker/src/ClayDatePicker.tsx

```diff
@@ -0,0 +1,379 @@
+	 */
+	years?: IYears;
+}
+
+const DateNow = new Date();
+
+/**
+ * ClayDatePicker component.
+ */
+const ClayDatePicker: FunctionComponent<Props> = ({
+	ariaLabels = {
+		buttonDot: 'Select current date',
+		buttonNextMonth: 'Select the next month',
+		buttonPreviousMonth: 'Select the previous month',
+	},
+	dateFormat = 'YYYY-MM-DD',
+	firstDayOfWeek = 0,
+	footerElement,
+	id,
+	initialExpanded = false,
+	initialMonth = DateNow,
+	inputName,
+	months = [
+		'January',
+		'February',
+		'March',
+		'April',
+		'May',
+		'June',
+		'July',
+		'August',
+		'September',
+		'October',
+		'November',
+		'December',
+	],
+	onValueChange = () => {},
+	placeholder,
+	spritemap,
+	time = false,
+	timeFormat = 'HH:mm',
+	timezone,
+	useNative = false,
+	value,
+	weekdaysShort = ['S', 'M', 'T', 'W', 'T', 'F', 'S'],
+	years = {
+		end: DateNow.getFullYear(),
+		start: DateNow.getFullYear(),
+	},
+}) => {
+	/**
+	 * Indicates the current month rendered on the screen.
+	 */
+	const [currentMonth, setCurrentMonth] = useCurrentMonth(initialMonth);
+
+	/**
+	 * Indicates the time selected by the user.
+	 */
+	const [currentTime, setCurrentTime] = useCurrentTime(timeFormat);
+
+	/**
+	 * The day selected by the user.
+	 */
+	const [daySelected, setDaySelected] = useState<Date>(initialMonth);
+
+	/**
+	 * An array of the weeks and days list for the current month
+	 */
+	const [weeks, setWeeks] = useWeeks(currentMonth, firstDayOfWeek);
+
+	/**
+	 * Flag to indicate if date is expanded.
+	 */
+	const [expanded, setExpanded] = useState(initialExpanded);
+
+	/**
+	 * Create a ref to store the datepicker DOM element
+	 */
+	const elementRef = useRef<HTMLDivElement | null>(null);
+
+	/**
+	 * Handles the click on element of the day
+	 */
+	function handleDayClicked(date: Date) {
+		setDaySelected(date);
+		onValueChange(date);
+	}
+
+	/**
+	 * Handles the input change.
+	 */
+	function inputChange(event: ChangeEvent<HTMLInputElement>) {
+		const {value} = event.target;
+		const format = time ? `${dateFormat} ${timeFormat}` : dateFormat;
+		const date = moment(value, format);
+		const year = date.year();
+
+		if (date.isValid() && year >= years.start && years.end >= year) {
+			setCurrentMonth(date.toDate());
+			setDaySelected(date.toDate());
+
+			if (time) {
+				setCurrentTime(date.hours(), date.minutes());
+			}
+		}
+
+		onValueChange(value);
+	}
+
+	/**
+	 * Handles dot cliecked
+	 */
+	const handleDotClicked = () => {
+		setCurrentMonth(initialMonth);
+		setDaySelected(initialMonth);
+		onValueChange(initialMonth);
+	};
+
+	/**
+	 * Handles datepicker view
+	 */
+	const handleCalendarButtonClicked = () => setExpanded(!expanded);
+
+	/**
+	 * Handles document click
+	 */
+	const handleDocClick = () => setExpanded(false);
+
+	useEffect(() => {
+		// It is not necessary to update the weeks when the native
+		// date is enabled. Unnecessary rendering and processing.
+		if (!useNative) {
+			setWeeks(currentMonth);
+		}
+	}, [currentMonth]);
```

Comment:
Why do we need to use `useEffect` for this? If I am reading this correctly, this will lead to a possible double render on first render and whenever the `currentMonth` changes and `!useNative` (because first we'll render/update, then `setWeeks` will update the state and we'll render a second time).

Reply:
We need `currentMonth` treated https://github.com/liferay/clay/pull/1440/files?file-filters%5B%5D=.ts#diff-1753466c95969933de66a3c9603a812bR31 before updating weeks. Maybe we can bring this method to DatePicker and treat the date before sending it to `setCurrentMonth` but we will have to call in several places and so we can update both at the same time and remove this `useEffect`, the problem is that I have to always make sure to call `setWeeks` when calling `setCurrentMonth`. What do you think?

Reply:
Makes sense. I think it's worth trying to refactor in order to get rid of `useEffect`; it seems important to get this pattern right early on because we wouldn't want double-rendering to be a thing we always do whenever we have dependent state.

Reply:
Yeah, I think we should start writing down these good practices on some of our style guides...